### DISCUSSION TITLE
Improve meson with clang toolchain on linux

### DISCUSSION THREAD

**Discussion by star-hengxing** [xmake-io/xmake#4979] [General]
linux + meson + clang 在链接的时候会调用 clang 进行链接，而不是 clang++，导致部分 c++ 库会缺 libstdc++ 符号。看了几个 meson issue，问题好像存在很久了，也影响了 xrepo 一些包：libsigcpp/libvips/openh264...

xmake 是否应该 workaround 一下，比如生成 native configs_file 这样可能可以解决？

---

linux + meson + clang calls clang for linking instead of clang++ when linking, causing some c++ libraries to be missing libstdc++ symbols. I've looked at a few meson issues, and it seems that the problem has existed for a long time, and has affected some xrepo packages: libsigcpp/libvips/openh264...

Should xmake workaround, e.g. generate native configs_file, which might solve it?

**Comment by waruqi** [comment]
目前已经是用了 native configs 了。https://github.com/xmake-io/xmake/blob/e39ff9d876e33cc36e4eb9e196ca665ecefc9f96/xmake/modules/package/tools/meson.lua#L337

可以调下然后来个 pr 改进下

**Comment by star-hengxing** [comment]
试了一下，要加个 `cpp_` 或 `c_` 前缀，但这个 build_getenv 似乎不能返回两个 toolset。

```lua
local ld = package:build_getenv("ld")
if ld then
    file:print("cpp_ld=['%s']", executable_path(ld))
end
```

另外还有一个问题，一些库会直接用 prefix 作为宏来使用，然后在 Windows 上因为路径问题编译失败。这个应该怎么改比较好
https://github.com/xmake-io/xmake-repo/actions/runs/8692471564/job/23837190006?pr=3466
https://github.com/xmake-io/xmake/blob/e39ff9d876e33cc36e4eb9e196ca665ecefc9f96/xmake/modules/package/tools/meson.lua#L298

**Comment by waruqi** [comment]
> 试了一下，要加个 `cpp_` 或 `c_` 前缀，但这个 build_getenv 似乎不能返回两个 toolset。
> 
> ```lua
> local ld = package:build_getenv("ld")
> if ld then
>     file:print("cpp_ld=['%s']", executable_path(ld))
> end
> ```

这样?

```lua
     file:print("cpp_ld=['%s']", executable_path(ld))
     file:print("c_ld=['%s']", executable_path(ld))
```


> 
> 另外还有一个问题，一些库会直接用 prefix 作为宏来使用，然后在 Windows 上因为路径问题编译失败。这个应该怎么改比较好 https://github.com/xmake-io/xmake-repo/actions/runs/8692471564/job/23837190006?pr=3466
> 
> https://github.com/xmake-io/xmake/blob/e39ff9d876e33cc36e4eb9e196ca665ecefc9f96/xmake/modules/package/tools/meson.lua#L298

没看到啥错误么

**Reply by star-hengxing** [reply]
> 没看到啥错误么

msvc 的 parser 报错总是比较诡异，直接列一下代码：

```c
#define VIPS_PREFIX "C:\Users\star\AppData\Local\Temp\.xmake\240416\xmake-repo\test\build\.packages\l\libvips\latest\76472218ac9c494e994db4613912f01a"

g_info("VIPS_PREFIX = %s", VIPS_PREFIX);
```

```
libvips/iofuncs/init.c(518): warning C4129: 's': unrecognized character escape sequence
libvips/iofuncs/init.c(518): warning C4129: 'A': unrecognized character escape sequence
libvips/iofuncs/init.c(518): warning C4129: 'L': unrecognized character escape sequence
libvips/iofuncs/init.c(518): warning C4129: 'T': unrecognized character escape sequence
libvips/iofuncs/init.c(518): warning C4129: '.': unrecognized character escape sequence
libvips/iofuncs/init.c(518): error C2153: integer literals must have at least one digit
libvips/iofuncs/init.c(518): warning C4129: 'l': unrecognized character escape sequence
libvips/iofuncs/init.c(518): error C2022: '500': too big for character
libvips/iofuncs/init.c(1047): warning C4129: 's': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): warning C4129: 'A': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): warning C4129: 'L': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): warning C4129: 'T': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): warning C4129: '.': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): error C2153: integer literals must have at least one digit
libvips/iofuncs/init.c(1047): warning C4129: 'l': unrecognized character escape sequence
libvips/iofuncs/init.c(1047): error C2022: '500': too big for character
libvips/iofuncs/init.c(1096): warning C4129: 's': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): warning C4129: 'A': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): warning C4129: 'L': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): warning C4129: 'T': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): warning C4129: '.': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): error C2153: integer literals must have at least one digit
libvips/iofuncs/init.c(1096): warning C4129: 'l': unrecognized character escape sequence
libvips/iofuncs/init.c(1096): error C2022: '500': too big for character
libvips/iofuncs/init.c(1184): warning C4129: 's': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): warning C4129: 'A': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): warning C4129: 'L': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): warning C4129: 'T': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): warning C4129: '.': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): error C2153: integer literals must have at least one digit
libvips/iofuncs/init.c(1184): warning C4129: 'l': unrecognized character escape sequence
libvips/iofuncs/init.c(1184): error C2022: '500': too big for character
```

**Reply by star-hengxing** [reply]
> ```lua
> file:print("cpp_ld=['%s']", executable_path(ld))
> file:print("c_ld=['%s']", executable_path(ld))
> ```

toolchain 是这样写的，为啥不能分开获取 c/cpp 的 toolset

```lua
set_toolset("ld", "clang++" .. suffix, "clang" .. suffix)
```

**Comment by waruqi** [comment]
> > ```lua
> > file:print("cpp_ld=['%s']", executable_path(ld))
> > file:print("c_ld=['%s']", executable_path(ld))
> > ```
> 
> toolchain 是这样写的，为啥不能分开获取 c/cpp 的 toolset
> 
> ```lua
> set_toolset("ld", "clang++" .. suffix, "clang" .. suffix)
> ```

目前不支持区分。。纯 c 也是用的 clang++ linker

**Comment by waruqi** [comment]
> > 没看到啥错误么
> 
> msvc 的 parser 报错总是比较诡异，直接列一下代码：
> 
> ```c
> #define VIPS_PREFIX "C:\Users\star\AppData\Local\Temp\.xmake\240416\xmake-repo\test\build\.packages\l\libvips\latest\76472218ac9c494e994db4613912f01a"
> 
> g_info("VIPS_PREFIX = %s", VIPS_PREFIX);
> ```
> 
> ```
> libvips/iofuncs/init.c(518): warning C4129: 's': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): warning C4129: 'A': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): warning C4129: 'L': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): warning C4129: 'T': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): warning C4129: '.': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): error C2153: integer literals must have at least one digit
> libvips/iofuncs/init.c(518): warning C4129: 'l': unrecognized character escape sequence
> libvips/iofuncs/init.c(518): error C2022: '500': too big for character
> libvips/iofuncs/init.c(1047): warning C4129: 's': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): warning C4129: 'A': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): warning C4129: 'L': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): warning C4129: 'T': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): warning C4129: '.': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): error C2153: integer literals must have at least one digit
> libvips/iofuncs/init.c(1047): warning C4129: 'l': unrecognized character escape sequence
> libvips/iofuncs/init.c(1047): error C2022: '500': too big for character
> libvips/iofuncs/init.c(1096): warning C4129: 's': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): warning C4129: 'A': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): warning C4129: 'L': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): warning C4129: 'T': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): warning C4129: '.': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): error C2153: integer literals must have at least one digit
> libvips/iofuncs/init.c(1096): warning C4129: 'l': unrecognized character escape sequence
> libvips/iofuncs/init.c(1096): error C2022: '500': too big for character
> libvips/iofuncs/init.c(1184): warning C4129: 's': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): warning C4129: 'A': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): warning C4129: 'L': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): warning C4129: 'T': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): warning C4129: '.': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): error C2153: integer literals must have at least one digit
> libvips/iofuncs/init.c(1184): warning C4129: 'l': unrecognized character escape sequence
> libvips/iofuncs/init.c(1184): error C2022: '500': too big for character
> ```

用 path.unix(prefix) 改成 unix path? `\\` -> `/`

**Reply by star-hengxing** [reply]
> 用 path.unix(prefix) 改成 unix path? \\ -> /

这个没问题。

但我觉得 Windows 传 installdir 之类都可以改成 unix path，或者输出的时候 `\` -> `\\`，不知道有没有潜在的风险，因为只有一些包需要这样手动 workaround。但外部应该是改不了 prefix 这个值，除非添加一个 opt.prefix

**Reply by waruqi** [reply]
tools.cmake 里面已经改了。。这种最好还是最小化改动比较好，遇到问题了，对应处理下就行。。可以规避 break 风险的同时，也能解决问题。。没必要一次性全改。。

**Reply by star-hengxing** [reply]
嗯，所以外部应该如何修改 prefix 呢，试了下 `package:set("installdir", path.unix(package:installdir()))` 不太行

**Reply by waruqi** [reply]
提个 pr 过来，直接 tools.meson 里面 --prefix 改下就行了

**Comment by star-hengxing** [comment]
破案了，这里返回的是 clang 而不是 clang++，导致链接的时候也是调 clang，在 Windows 上也是同样的行为

https://github.com/xmake-io/xmake/blob/e39ff9d876e33cc36e4eb9e196ca665ecefc9f96/xmake/modules/package/tools/meson.lua#L213-L216

**Reply by waruqi** [reply]
可以参考这里改下 https://github.com/xmake-io/xmake/blob/e39ff9d876e33cc36e4eb9e196ca665ecefc9f96/xmake/modules/package/tools/cmake.lua#L546
